Set the local and remote CIDRs for the aws_vpn_connection resource

jsf9k opened the pull request:
## 🗣 Description ##

Set the local and remote CIDRs for the `aws_vpn_connection` resource via the `local_ipv4_network_cidr` and `remote_ipv4_network_cidr` keys.

This pull request also changes `versions.tf` to require version 3.43.0 or greater of the AWS Terraform provider, since version 3.43.0 is the first version that fixes a bug where these CIDRs were incorrectly required to be `/32`s.

## 💭 Motivation and context ##

I had been working around the bug in Terraform AWS provider by applying _without_ specifying values for the `local_ipv4_network_cidr` and `remote_ipv4_network_cidr` keys, then going into the AWS console and manually setting them to the correct values.  With this fix such manual intervention is no longer necessary.

See hashicorp/terraform-provider-aws#16879 and hashicorp/terraform-provider-aws#17573 for more details about the bugfix in the Terraform AWS provider.

## 🧪 Testing ##

I applied these changes to our production COOL environment (where I had already done the manual intervention described in the previous section) and verified that the `terraform apply` did not want to make changes to any resources.

## ✅ Checklist ##

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
* [x] All new and existing tests pass.

jsf9k committed:
Set the local and remote CIDRs for the aws_vpn_connection resource

Also require version 3.43.0 or greater of the AWS Terraform provider,
since that is the first version that fixes a bug where these CIDRs
were incorrectly required to be /32s.

jsf9k committed:
Set the remote network CIDR to only the bottom /20 of Shared Services

This is because we want to give CDM as small a subnet as possible,
since they are assuming there are no duplicates across the CISA
enterprise.  All the instances that interest them (the OpenVPN and
FreeIPA instances) are in the bottom /4 of the SharedServices CIDR
block.

jsf9k committed:
Fix incorrect provider versions in README

dav3r reviewed (approved):
LGTM - congrats on getting your [Terraform PR](https://github.com/hashicorp/terraform-provider-aws/pull/17573) merged! 🎉

jsf9k commented:
> LGTM - congrats on getting your [Terraform PR](https://github.com/hashicorp/terraform-provider-aws/pull/17573) merged!

"Made it, Ma! Top of the world!"

felddy commented on s2s_vpn_tunnel.tf:
```diff
@@ -19,20 +19,15 @@ resource "aws_customer_gateway" "cdm" {
+  # duplicates across the CISA enterprise.  All the instances that
+  # interest them (the OpenVPN and FreeIPA instances) are in the
+  # bottom /4 of the SharedServices CIDR block.
+  remote_ipv4_network_cidr = cidrsubnets(data.terraform_remote_state.networking.outputs.vpc.cidr_block, 4)[0]
```
Magic `4`.

mcdonnnj replied:
The way this is phrased (and coded) makes it sound like the value of `data.terraform_remote_state.networking.outputs.vpc.cidr_block` is a `/24` and you're subnetting it down to a `/28` to get a resultant subnet for the bottom `/4`. If that's correct I think it would be beneficial to elaborate on what is being done more in the comments so it is clearer what is happening as I believe that value is set by a variable in [cisagov/cool-sharedservices-networking/vpc.tf](https://github.com/cisagov/cool-sharedservices-networking/blob/5a849d960d25fdf1cfab7a2788023f7a58d14420/vpc.tf#L17). If what I'm saying is wrong then I would like an explanation even more.

felddy reviewed (approved):
Fantabulous.  And congrats!  🎉 
@jsf9k right now: 👇 
![magnum](https://user-images.githubusercontent.com/3229435/120383210-9bff3500-c2f2-11eb-95fc-cdffebb300da.gif)

jsf9k committed:
Correct version constraint

The version constraint ~> x.y actually matches x.y (or x.y.0).

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj reviewed (changes requested):
This change makes sense and thanks for putting in the research and following progress in the AWS provider. I would like some commenting to more explicitly spell out what is being done since it interacts with a remote state that is (I believe) in turn affected by variable input.

jsf9k committed:
Improve/correct comment describing subnet calculation

Also use cidrsubnet() instead of cidrsubnets() since we only need the
first subnet.

Thanks to @mcdonnnj for pointing out that my previous explanation was
incorrect and unclear.

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k commented:
> This change makes sense and thanks for putting in the research and following progress in the AWS provider. I would like some commenting to more explicitly spell out what is being done since it interacts with a remote state that is (I believe) in turn affected by variable input.

You're right, my previous explanation was both incorrect and unclear.  Hopefully I have improved it in commit ef94d35.

dav3r commented on s2s_vpn_tunnel.tf:
```diff
@@ -19,20 +19,22 @@ resource "aws_customer_gateway" "cdm" {
-  # https://github.com/hashicorp/terraform-provider-aws/issues/16879
+  customer_gateway_id     = aws_customer_gateway.cdm.id
+  local_ipv4_network_cidr = var.cdm_cidr
+  # The cidrsubnets() here is because we want to give CDM as small a
```
Singular now, yes?
```suggestion
  # The cidrsubnet() here is because we want to give CDM as small a
```

jsf9k committed:
Correct function name in comment

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

dav3r reviewed (approved):
Double-approved! 👍 👍

mcdonnnj reviewed (approved):
The new comment is
![image](https://user-images.githubusercontent.com/50747025/120810941-7bfd8a80-c519-11eb-84aa-10d20d50ca32.png)

Thanks for thoroughly explaining the logic of what's being done here.
